keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,345 @@
+
+```
+                  +-----------------+-------------------------+
+NUMA node0(40Gi)  | Safety Zone(5Gi)|    Allocatable(35Gi)    |
+                  +-----------------+-------------------------+
+                  +-----------------+-------------------------+
+NUMA node1(40Gi)  | Safety Zone(5Gi)|    Allocatable(35Gi)    |
+                  +-----------------+-------------------------+
+```
+_Figure: The example of safety zone._
+
+
+#### The Concept of Reserved Zone
+
+Memory Manager has the concept of reserved zone to calculate NUMA node affinity precisely. The reserved zone is literally reserved for the containers which has Guaranteed QoS. Container doesn't always consume memory as much as they requested. For example container may request 5Gbi memory, but it can consumes only 3Gbi at runtime. Memory Manager treats requested memory of Guaranteed QoS containers as reserved memory. Memory Manager guarantee Guaranteed QoS containers usage of memory and hugepages on single NUMA node as they requested by using [CPUSET enforcement](#cpuset-enforcement) and [OOM Score Enforcement](#oom-score-enforcement). 
+
+```
+           +---------------+------------+--------------+
+           |               |       Allocatable         |
+NUMA node  +  Safety Zone  +------------+--------------+
+           |               |  Reserved  |     free     |
+           +---------------+------------+--------------+
+```
+_Figure: The memory map of NUMA node._
+
+NOTE:
+- Reserved zone cannot be greater than Allocatable zone.
+- Typical Linux processes and containers which have BestEffort or Burstable QoS may consume memory on a NUMA node where Guaranteed QoS containers also consumes memory. It is also possible that Guaranteed QoS containers requested lots of memory but consumes just a little bit of them and the others occupy reserved zone when only few free memory is available on the NUMA node. At this situation, OOM may happen by Guaranteed QoS Container if the container try to consume more memory. Then others will be killed by OOM killer to make free memory. Because, unlike Guaranteed QoS containers, the others has high OOM socre(which has low priority at OOM system).
+
+#### NUMA node affinity for memory and hugepages
+
+Memory Manager calculates NUMA node affinity that represents which NUMA node has enough capacity of memory and/or hugepages for containers in Guaranteed QoS Pod. To calculate affinity, Memory Manager takes capacity of memory and pre-allocated hugepages per NUMA node on its intialize sequence then calculate `Allocatable Memory of NUMA node` for every NUMA nodes.  Then when pod admission is requested, Memory Manager checks the pod has Guaranteed QoS or not, if the pod has Guaranteed QoS, Memory Manager calculates NUMA node affinity by the following way: it checks resources availability of each NUMA nodes based on `Allocatable Memory of NUMA node` and total amount of `reserved` memory and hugepages of NUMA node.
```

Comment:
I assume this calculation will be done for all valid hugepage sizes?

Reply:
https://github.com/kubernetes/enhancements/pull/1203/files#r376887295
sure :), as I commented on the above,